File: web/lite/src/base/video-composition.ts
import { Composition } from './composition';
import { CompositionType } from '../codec/types';
import { TimeRange } from './time-range';
import { VideoSequence } from './video-sequence';
import { verifyFailed } from './utils/verify';

export class VideoComposition extends Composition {
  public hasAlpha = false;
  public sequences: Array<VideoSequence> = [];

  private staticTimeRanges: Array<TimeRange> = [];
  private staticTimeRangeUpdated = false;

  public type(): CompositionType {
    return CompositionType.Video;
  }

  public getStaticTimeRanges(): Array<TimeRange> {
    if (!this.staticTimeRangeUpdated) {
      this.staticTimeRangeUpdated = true;
      this.updateStaticTimeRanges();
    }
    return this.staticTimeRanges;
  }

  public updateStaticTimeRanges(): void {
    if (this.duration <= 1) return;
    if (this.sequences.length > 0) {
      let sequence = this.sequences[0];
      for (let i = 1; i < this.sequences.length; i++) {
        const item = this.sequences[i];
        if (item.frameRate > sequence.frameRate) sequence = item;
      }
      const timeScale = this.frameRate / sequence.frameRate;
      for (const timeRange of sequence.staticTimeRanges) {
        timeRange.start = Math.round(timeRange.start * timeScale);
        timeRange.end = Math.round(timeRange.end * timeScale);
        this.staticTimeRanges.push(timeRange);
      }
    } else {
      const range: TimeRange = { start: 0, end: this.duration - 1 };
      this.staticTimeRanges.push(range);
    }
  }

  public hasImageContent(): boolean {
    return true;
  }

  public verify(): boolean {
    if (!super.verify() || this.sequences.length <= 0) {
      verifyFailed();
      return false;
    }
    for (const sequence of this.sequences) {
      if (!sequence || !sequence.verify()) {
        verifyFailed();
        return false;
      }
    }
    return true;
  }
}

File: src/base/VideoComposition.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////
//
//  Tencent is pleased to support the open source community by making libpag available.
//
//  Copyright (C) 2021 Tencent. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  unless required by applicable law or agreed to in writing, software distributed under the
//  license is distributed on an "as is" basis, without warranties or conditions of any kind,
//  either express or implied. see the license for the specific language governing permissions
//  and limitations under the license.
//
/////////////////////////////////////////////////////////////////////////////////////////////////

#include "base/utils/Verify.h"
#include "pag/file.h"

namespace pag {
VideoComposition::~VideoComposition() {
  for (auto sequence : sequences) {
    delete sequence;
  }
}

CompositionType VideoComposition::type() const {
  return CompositionType::Video;
}

void VideoComposition::updateStaticTimeRanges() {
  staticTimeRanges = {};
  if (duration <= 1) {
    return;
  }
  if (!sequences.empty()) {
    auto sequence = sequences[0];
    for (size_t i = 1; i < sequences.size(); i++) {
      auto item = sequences[i];
      if (item->frameRate > sequence->frameRate) {
        sequence = item;
      }
    }
    float timeScale = frameRate / sequence->frameRate;
    for (auto timeRange : sequence->staticTimeRanges) {
      timeRange.start = static_cast<Frame>(roundf(timeRange.start * timeScale));
      timeRange.end = static_cast<Frame>(roundf(timeRange.end * timeScale));
      staticTimeRanges.push_back(timeRange);
    }
  } else {
    TimeRange range = {0, duration - 1};
    staticTimeRanges.push_back(range);
  }
}

bool VideoComposition::hasImageContent() const {
  return true;
}

bool VideoComposition::verify() const {
  if (!Composition::verify() || sequences.empty()) {
    VerifyFailed();
    return false;
  }
  auto sequenceValid = [](VideoSequence* sequence) {
    return sequence != nullptr && sequence->verify();
  };
  if (!std::all_of(sequences.begin(), sequences.end(), sequenceValid)) {
    VerifyFailed();
    return false;
  }
  return true;
}
}  // namespace pag
